File: examples/nodejs/online_smoothing.js
const fastlowess = require('../../bindings/nodejs');

/**
 * fastlowess Online Smoothing Example
 * 
 * This example demonstrates online LOWESS smoothing for real-time data:
 * - Basic incremental processing with streaming data
 * - Real-time sensor data smoothing
 * - Different update modes (Full vs Incremental)
 * - Memory-bounded processing with sliding window
 */

function main() {
    console.log("=== fastlowess Online Smoothing Example ===");

    // 1. Simulate a real-time signal
    // A sine wave with changing frequency and random noise
    const nPoints = 1000;
    const x = new Float64Array(nPoints);
    const yTrue = new Float64Array(nPoints);
    const y = new Float64Array(nPoints);

    for (let i = 0; i < nPoints; i++) {
        x[i] = i;
        yTrue[i] = 20.0 + 5.0 * Math.sin(x[i] * 0.1) + 2.0 * Math.sin(x[i] * 0.02);
        // Gaussian noise
        let u1 = Math.random();
        let u2 = Math.random();
        let z = Math.sqrt(-2.0 * Math.log(u1)) * Math.cos(2.0 * Math.PI * u2);
        y[i] = yTrue[i] + z * 1.2;
    }

    // Add some sudden spikes (sensor glitches)
    for (let i = 200; i < 205; i++) y[i] += 15.0;
    for (let i = 600; i < 610; i++) y[i] -= 10.0;

    console.log(`Simulating ${nPoints} real-time data points...`);

    // 2. Sequential Online Processing
    
    // Full Update Mode (higher accuracy)
    console.log("Processing with 'full' update mode...");
    const onlineFull = new fastlowess.OnlineLowess(
        { fraction: 0.3, iterations: 3 },
        { windowCapacity: 50, updateMode: "full" }
    );
    const resFull = new Float64Array(nPoints);
    for (let i = 0; i < nPoints; i++) {
        const chunkX = new Float64Array([x[i]]);
        const chunkY = new Float64Array([y[i]]);
        const res = onlineFull.addPoints(chunkX, chunkY);
        resFull[i] = res.y[0];
    }

    // Incremental Update Mode (faster for large windows)
    console.log("Processing with 'incremental' update mode...");
    const onlineInc = new fastlowess.OnlineLowess(
        { fraction: 0.3, iterations: 3 },
        { windowCapacity: 50, updateMode: "incremental" }
    );
    const resInc = new Float64Array(nPoints);
    for (let i = 0; i < nPoints; i++) {
        const chunkX = new Float64Array([x[i]]);
        const chunkY = new Float64Array([y[i]]);
        const res = onlineInc.addPoints(chunkX, chunkY);
        resInc[i] = res.y[0];
    }

    // Compare results
    console.log("\nResults Comparison:");

    // Show sample around spike area
    console.log("\nSample around spike (indices 198-208):");
    console.log("Index\tRaw\t\tTrue\t\tFull\t\tIncremental");
    for (let i = 198; i <= 208; i++) {
        // Handle initial points where output might be null (though by 198 it should be fine)
        const fullVal = isNaN(resFull[i]) ? "NaN" : resFull[i].toFixed(2);
        const incVal = isNaN(resInc[i]) ? "NaN" : resInc[i].toFixed(2);
        console.log(`${i}\t${y[i].toFixed(2)}\t\t${yTrue[i].toFixed(2)}\t\t${fullVal}\t\t${incVal}`);
    }

    // Calculate overall statistics
    // Filter out initial NaNs
    let mseFull = 0;
    let mseInc = 0;
    let count = 0;

    for (let i = 0; i < nPoints; i++) {
        if (!isNaN(resFull[i]) && !isNaN(resInc[i])) {
            mseFull += Math.pow(resFull[i] - yTrue[i], 2);
            mseInc += Math.pow(resInc[i] - yTrue[i], 2);
            count++;
        }
    }
    mseFull /= count;
    mseInc /= count;

    console.log("\nMean Squared Error vs True Signal:");
    console.log(` - Full Update:        ${mseFull.toFixed(4)}`);
    console.log(` - Incremental Update: ${mseInc.toFixed(4)}`);

    console.log("\n=== Online Smoothing Example Complete ===");
}

main();

File: examples/julia/online_smoothing.jl
#!/usr/bin/env julia
"""
FastLOWESS Online Smoothing Example

This example demonstrates online LOWESS smoothing for real-time data:
- Basic incremental processing with streaming data
- Real-time sensor data smoothing
- Different update modes (Full vs Incremental)
- Memory-bounded processing with sliding window

The OnlineLowess class is designed for:
- Real-time data streams
- Sensors and monitoring
- Low-latency applications
"""

using Random
using Printf

# Handle package loading - check if we're already in the FastLOWESS project
using Pkg
project_name = Pkg.project().name
if project_name != "FastLOWESS"
    # Not in the FastLOWESS project, need to develop it
    script_dir = @__DIR__
    julia_pkg_dir = joinpath(dirname(script_dir), "julia")
    if !haskey(Pkg.project().dependencies, "FastLOWESS")
        Pkg.develop(path = julia_pkg_dir)
    end
end

using FastLOWESS

function main()
    println("=== FastLOWESS Online Smoothing Example ===")

    # 1. Simulate a real-time signal
    # A sine wave with changing frequency and random noise
    n_points = 1000
    Random.seed!(42)
    x = collect(Float64, 0:(n_points-1))
    y_true = 20.0 .+ 5.0 .* sin.(x .* 0.1) .+ 2.0 .* sin.(x .* 0.02)
    y = y_true .+ randn(n_points) .* 1.2

    # Add some sudden spikes (sensor glitches)
    y[201:205] .+= 15.0
    y[601:610] .-= 10.0

    println("Simulating $n_points real-time data points...")

    # 2. Sequential Online Processing
    # Full Update Mode (higher accuracy)
    println("Processing with 'full' update mode...")
    model_full = OnlineLowess(
        fraction = 0.3,
        window_capacity = 50,
        iterations = 3,
        update_mode = "full",
    )
    res_full = add_points(model_full, x, y)

    # Incremental Update Mode (faster for large windows)
    println("Processing with 'incremental' update mode...")
    model_inc = OnlineLowess(
        fraction = 0.3,
        window_capacity = 50,
        iterations = 3,
        update_mode = "incremental",
    )
    res_inc = add_points(model_inc, x, y)

    # Compare results
    println("\nResults Comparison:")

    # Show sample around spike area
    println("\nSample around spike (indices 198-208):")
    println("Index\tRaw\t\tTrue\t\tFull\t\tIncremental")
    for i = 198:208
        @printf(
            "%d\t%.2f\t\t%.2f\t\t%.2f\t\t%.2f\n",
            i,
            y[i],
            y_true[i],
            res_full.y[i],
            res_inc.y[i]
        )
    end

    # Calculate overall statistics
    mse_full = sum((res_full.y .- y_true) .^ 2) / n_points
    mse_inc = sum((res_inc.y .- y_true) .^ 2) / n_points

    println("\nMean Squared Error vs True Signal:")
    @printf(" - Full Update:        %.4f\n", mse_full)
    @printf(" - Incremental Update: %.4f\n", mse_inc)

    println("\n=== Online Smoothing Example Complete ===")
end

main()
